design-documents/simplified-module-declaration.md

```diff
@@ -2,39 +2,43 @@
-* Eliminate `module.xml` - composer.json contains 'real' module name. Move 'legacy' module names in `composer.json`
-* Use `composer.json` for module declaration. 
-
-## Open Questions
-
-As described by Vinai Kopp, performance of composer dependency resolution impacts developer experience. 
-
-`registration.php` file allows to quickly install a module without composer. Moving to composer will remove this optimisation. 
-
-The solution to this problem should be provided before `registration.php` can be removed. Possible options:
-
-* make `registration.php` optional instead of getting rid of it
-* add command to composer that will allow to re-read composer.json files of magento modules without full dependency resolution
+Magento module metadata is declared in the following files today:
+
+* `module.xml` - declares module name and configuration load order dependencies
+* `registration.php` - tells Magento where the module resides for Magento to be able to read the module configuration files
+* (optional) `composer.json` - declares package name, description, version, dependencies, etc. It is optional and only becomes required if you want to distribute your module through Composer.
+
+This implementation of Magento module declaration leads to some inconveniences:
+* the standard PHP way of declaring packages (`composer.json`) is optional if you want to write a module and becomes required if you're going to distribute it
+* Magento has its custom package declaration format that partly duplicates the `composer.json` format
+* every `registration.php` is included on every request. OPCache reduces the performance impact. But the whole problem can be avoided
+
+## Proposal
+
+* Make `composer.json` the default way of declaring modules in Magento. `composer.json` is the standard package declaration format in PHP world. Switching to `composer.json` as the module declaration file will make Magento modules more familiar for non-Magento PHP developers and will reduce the need for duplication.
+  * Modules in `/app/code` will still be supported. They will not be installed using Composer, but they will declare their metadata in `composer.json`
+* Make `registration.php` optional
+  * Add Composer installation hook that will:
+    * Find all Composer packages installed in `/vendor` with `"type": "magento2-*"`
+    * Find all folders in `app/code` containing `composer.json` with `"type": "magento2-*"`
+    * Merge the lists and sort components based on their dependencies declared in `require` section of `composer.json` and `sequence` section of `module.xml`
+    * Write the sorted list to `/vendor/magento_components.php`
+  * This will reduce the number of files included by PHP process on every request
+* Make `module.xml` optional. `composer.json` contains 'real' module name.
+  * Determining module name
+    * if only a composer.json is present, determine the module name from the package name, e.g. magento/module-hello-world becomes Magento_HelloWorld.
```

Comment:
Is the Vendor_ModuleName format really necessary? Why not take this as an opportunity to drop it?

Reply:
I don't see the harm in it.
And it's currently being used the `setup_module` table in the database to keep track of db migration versions (but one day that's going to disappear as well I guess?)

Reply:
@navarr Dropping it would break all existing modules.  

&lt;sarcasm>Everybody loves paying for hours "fixing" working code only to work around backward incompatible changes that add no value at all.&lt;/sarcasm>

Backward compatibility should be prioritized over "cleaning up" (at least in my book).

Reply:
I don't believe we'd have to break all existing modules to deprecate an old syntax and smoothly transition to a new one.

The biggest case of backwards incompatibility here would likely be version identifiers from pre-schema transitions.  A similar issue that occurs if the auto-generated name doesn't match the composer name for a module that wants to drop `module.xml`

This would be a good opportunity to begin such a cleanup/transition.

Reply:
It would break all the code that expects the old format. This includes not only custom modules but also external tooling.
Can you elaborate the benefits you see?

Reply:
There are a lot of places in Magento core which expect this, all the way down to fancy regex expressions used to validate layout XML where module names are used heavily in reference to template files. Agree with Vinai, this name format should be kept as the alternative is less than acceptable in my book.